crates/fuel-core/src/graphql_api/da_compression.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{
+    fuel_core_graphql_api::ports::worker::OffChainDatabaseTransaction,
+    graphql_api::storage::da_compression::{
+        DaCompressedBlocks,
+        DaCompressionTemporalRegistry,
+        DaCompressionTemporalRegistryIndex,
+    },
+};
+use fuel_core_compression::{
+    ports::{
+        TemporalRegistry,
+        UtxoIdToPointer,
+    },
+    services::compress::compress,
+    RegistryKeyspace,
+};
+use fuel_core_storage::{
+    not_found,
+    StorageAsMut,
+    StorageAsRef,
+    StorageInspect,
+};
+use fuel_core_types::{
+    blockchain::block::Block,
+    services::executor::Event,
+};
+use futures::FutureExt;
+
+/// Performs DA compression for a block and stores it in the database.
+pub fn da_compress_block<T>(
+    block: &Block,
+    events: &[Event],
+    transaction: &mut T,
+) -> anyhow::Result<()>
+where
+    T: OffChainDatabaseTransaction,
+    T: StorageInspect<DaCompressionTemporalRegistry>,
+{
+    let compressed = compress(CompressTx(transaction, events), block)
+        .now_or_never()
+        .expect("The current implementation resolved all futures instantly")?;
+
+    transaction
+        .storage_as_mut::<DaCompressedBlocks>()
+        .insert(&block.header().consensus().height, &compressed)?;
+
+    Ok(())
+}
+
+struct CompressTx<'a, Tx>(&'a mut Tx, &'a [Event]);
+
+impl<'a, Tx> TemporalRegistry for CompressTx<'a, Tx>
+where
+    Tx: OffChainDatabaseTransaction,
+    Tx: StorageInspect<DaCompressionTemporalRegistry>,
+{
+    fn read_registry(
+        &self,
+        keyspace: RegistryKeyspace,
+        key: fuel_core_types::fuel_compression::RegistryKey,
+    ) -> anyhow::Result<Vec<u8>> {
+        Ok(self
+            .0
+            .storage_as_ref::<DaCompressionTemporalRegistry>()
+            .get(&(keyspace, key))?
+            .ok_or(not_found!(DaCompressionTemporalRegistry))?
+            .into_owned())
+    }
+
+    fn write_registry(
+        &mut self,
+        keyspace: RegistryKeyspace,
+        key: fuel_core_types::fuel_compression::RegistryKey,
+        value: Vec<u8>,
+    ) -> anyhow::Result<()> {
+        // Write the actual value
+        self.0
+            .storage_as_mut::<DaCompressionTemporalRegistry>()
+            .insert(&(keyspace, key), &value)?;
+
+        // Remove the overwritten value from index, if any
+        self.0
+            .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+            .remove(&(keyspace, value.clone()))?;
+
+        // Add the new value to the index
+        self.0
+            .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+            .insert(&(keyspace, value), &key)?;
```

Comment:
is there a way we can make this group of operations atomic?

Reply:
They are atomic, as they operate on an atomic transaction